### PR TITLE
[CALCITE-2965] Implement string functions: REPEAT, SPACE, SOUNDEX, DI…

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-2965

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -243,6 +249,49 @@ public static int ascii(String s) {
...
+  public static String repeat(String s, int n) {
+    if (n < 1) {
+      return "";
+    }
+    int len = s.length() * n;
+    StringBuilder buf = new StringBuilder(len);
+    for (int i = 0; i < n; i++) {
+      buf.append(s);
+    }
+    return buf.toString();
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: from L257~262, you can use Strings.repeat

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -243,6 +249,49 @@ public static int ascii(String s) {
...
+  public static String space(int n) {
+    if (n < 1) {
+      return "";
+    }
+
+    StringBuilder buf = new StringBuilder(n);
+    for (int i = 0; i < n; i++) {
+      buf.append(" ");
+    }
+    return buf.toString();
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: we can just call repeat(" ", n)?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 1 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Will do.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Will do.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 3]
```diff
@@ -243,6 +250,36 @@ public static int ascii(String s) {
...
+  public static String space(int n) {
+    return repeat(" ", n);
+  }
+
+  /** SQL SOUNDEX(STRING) function. */
+  public static String soundex(String s) {
+    return SOUNDEX.soundex(s);
+  }
+
+  /** SQL DIFFERENCE(STRING, STRING) function. */
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: string in lower case.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 4]
```diff
@@ -243,6 +250,36 @@ public static int ascii(String s) {
...
+    }
+    return Strings.repeat(s, n);
+  }
+
+  /** SQL SPACE(int) function. */
+  public static String space(int n) {
+    return repeat(" ", n);
+  }
+
+  /** SQL SOUNDEX(STRING) function. */
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: string in lower case.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 3 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 4 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done.

**Comment by chunweilei** [comment]
@hsyuan Updated. Thank you for your review. But I am afraid that it should not be merged until we come to an agreement on where to put these functions which is discussed in [CALCITE-2965](https://issues.apache.org/jira/browse/CALCITE-2965).

**Comment by chunweilei** [comment]
Rebase the PR to resolve confilcts.

**Comment by hsyuan** [comment]
LGTM.

**Comment by chunweilei** [comment]
Update to fix build fail.

**Comment by chunweilei** [comment]
Update to resolve conflicts.

**Pull request merged**